packages/db/src/query/reports.ts

```diff
@@ -0,0 +1,167 @@
+                and(
+                    eq(sql`DATEDIFF(NOW(), reports.timestamp_last)`, reportConfig.interval),
+                    lte(reportConfig.time, new Date().getHours()),
+                ),
+            ),
+        );
+}
+
+export async function getLastReportForUser(userId: string): Promise<ReportProps | null> {
+    const userReports = await getReportForUserDescending(userId);
+
+    if (!userReports || userReports.length === 0) {
+        console.info("No report found for user");
+        return null;
+    }
+
+    const lastReport = userReports[0];
+
+    return {
+        dateFrom: new Date(lastReport.dateFrom).toString(),
+        dateTo: new Date(lastReport.dateTo).toString(),
+        totalEnergyConsumption: lastReport.totalEnergyConsumption,
+        avgEnergyConsumptionPerDay: lastReport.avgEnergyConsumptionPerDay,
+        totalEnergyCost: lastReport.totalEnergyCost,
+        avgEnergyCost: lastReport.avgEnergyCost,
+        highestPeak: {
+            dateTime: (lastReport.highestPeakDateTime ?? new Date(0)).toLocaleDateString(),
+            deviceName: lastReport.highestPeakDeviceName ?? "",
+            consumption: lastReport.highestPeakConsumption ?? "",
+        },
+        dayEnergyStatistics: [],
+    };
+}
+
+export async function getReportForUserDescending(userId: string) {
+    return db
+        .select({
+            id: reports.id,
+            dateFrom: reports.dateFrom,
+            dateTo: reports.dateTo,
+            totalEnergyConsumption: reports.totalEnergyConsumption,
+            avgEnergyConsumptionPerDay: reports.avgEnergyConsumptionPerDay,
+            totalEnergyCost: reports.totalEnergyCost,
+            avgEnergyCost: reports.avgEnergyCost,
+            highestPeakDateTime: reports.highestPeakDateTime,
+            highestPeakDeviceName: reports.highestPeakDeviceName,
+            highestPeakConsumption: reports.highestPeakConsumption,
+        })
+        .from(reports)
+        .where(eq(reports.userId, userId))
+        .orderBy(desc(reports.timestamp));
+}
+
+export async function saveReport(reportProps: ReportProps, userId: string) {
+    return db.transaction(async (trx) => {
+        await trx.insert(reports).values({
+            timestamp: new Date(),
+            dateFrom: new Date(reportProps.dateFrom),
+            dateTo: new Date(reportProps.dateTo),
+            userId: userId,
+            totalEnergyConsumption: reportProps.totalEnergyConsumption,
+            avgEnergyConsumptionPerDay: reportProps.avgEnergyConsumptionPerDay,
+            totalEnergyCost: reportProps.totalEnergyCost,
+            avgEnergyCost: reportProps.avgEnergyCost,
+            highestPeakDateTime: new Date(reportProps.highestPeak.dateTime),
+            highestPeakDeviceName: reportProps.highestPeak.deviceName,
+            highestPeakConsumption: reportProps.highestPeak.consumption,
+        });
+        const newReport = await trx
```

Comment:
Du könntest auch die ID oben selbst erzeugen und in den Insert reingeben. Dann sparst du dir hier den Select

Reply:
Das ist der bevorzugte Weg

Reply:
Zum Verständnis: Ist der bevorzugte Weg so, wie ich es vorgeschlagen habe, oder so, wie es bereits umgesetzt ist?

Reply:
Wie du es Vorgeschlagen hast